runtime/src/main/cpp/Memory.cpp

```diff
@@ -2262,44 +2266,48 @@ void freezeAcyclic(ContainerHeader* rootContainer, ContainerHeaderSet* newlyFroz
   }
 }
 
-void freezeCyclic(ContainerHeader* rootContainer,
+void freezeCyclic(ObjHeader* root,
                   const KStdVector<ContainerHeader*>& order,
                   ContainerHeaderSet* newlyFrozen) {
   KStdUnorderedMap<ContainerHeader*, KStdVector<ContainerHeader*>> reversedEdges;
-  KStdDeque<ContainerHeader*> queue;
-  queue.push_back(rootContainer);
+  KStdDeque<ObjHeader*> queue;
+  queue.push_back(root);
   while (!queue.empty()) {
-    ContainerHeader* current = queue.front();
+    ObjHeader* current = queue.front();
     queue.pop_front();
-    current->unMark();
-    reversedEdges.emplace(current, KStdVector<ContainerHeader*>(0));
-    traverseContainerReferredObjects(current, [current, &queue, &reversedEdges](ObjHeader* obj) {
+    ContainerHeader* currentContainer = current->container();
+    currentContainer->unMark();
+    reversedEdges.emplace(currentContainer, KStdVector<ContainerHeader*>(0));
+    traverseContainerReferredObjects(currentContainer, [current, currentContainer, &queue, &reversedEdges](ObjHeader* obj) {
           ContainerHeader* objContainer = obj->container();
           if (canFreeze(objContainer)) {
             if (objContainer->marked())
-              queue.push_back(objContainer);
-            reversedEdges.emplace(objContainer, KStdVector<ContainerHeader*>(0)).first->second.push_back(current);
+              queue.push_back(obj);
+            // We ignore FreezableAtomicsReference during condensation, to avoid KT-33824.
```

Comment:
`order` then is likely: `node` `ref`.

The condensation loop below then processes `ref` first, traverses its **reverse** edges and finds `node` a part of the same SCC as `ref`.

Reply:
Please provide a test where it actually happens. Previous algorithm just got bug accounting for inner references.

Reply:
Incorrect condensation here _actually_ happens.
This can be detected if the test above is compiled in debug mode. Also becomes visible with memory tracing.

Also can be observed through weak reference (as leaking `node`):
```kotlin
import kotlin.native.concurrent.*
import kotlin.native.ref.*

class Node(val next: Any?)

lateinit var weakRef: WeakReference<Any>

fun createRef(): FreezableAtomicReference<Any?> {
    val ref = FreezableAtomicReference<Any?>(null)
    val node = Node(ref)
    weakRef = WeakReference(node)
    ref.value = node
    ref.freeze()
    return ref
}

fun main() {
    val ref = createRef()
    ref.value = null
    kotlin.native.internal.GC.collect()

    println(weakRef.get())
} 
```

Reply:
Indeed, we shall ignore FreezableAtomic references in both directions when building condensation, fixing and adding above tests.

Reply:
> we shall ignore FreezableAtomic references in both directions when building condensation

The patch you propose in https://github.com/JetBrains/kotlin-native/pull/3337/commits/344dcd833c0c792594c3ba46bed50b7336cfb20e doesn't actually fix the root cause that I mentioned above:

> Ignoring these edges only here isn't enough, since they have already affected `order` making the entire SCC building algorithm incorrect.

Incorrect condensation can be found in the following test:
```kotlin
import kotlin.native.concurrent.*
import kotlin.native.ref.*

class Node(val next: Any?)

lateinit var node1: Node
lateinit var weakNode2: WeakReference<Node>

fun createRef() {
    val ref = FreezableAtomicReference<Any?>(null)
    node1 = Node(ref)
    val node2 = Node(node1)
    weakNode2 = WeakReference(node2)
    ref.value = node2

    node1.freeze()
    ref.value = null
}

fun main() {
    createRef()
    kotlin.native.internal.GC.collect()
    println(weakNode2.get())
}
``` 
Here `node1` and `node2` are found to form an SCC, which is incorrect.
This can be observed either through the weak reference or with memory tracing.

Reply:
Rewritten